internal/server/config.go

```diff
@@ -523,56 +523,88 @@ func getKindFromUnstructured(data interface{}) string {
 	return ""
 }
 
-func (s Server) loadConfig(config Config) error {
-	if err := validator.New().Struct(config); err != nil {
-		return err
-	}
-
+func (s Server) loadConfig() error {
 	return s.db.Transaction(func(tx *gorm.DB) error {
-		// inject internal infra provider
-		config.Providers = append(config.Providers, Provider{
-			Name: models.InternalInfraProviderName,
-		})
+		// load default infra provider
+		provider, err := s.loadProvider(tx, Provider{Name: models.InternalInfraProviderName})
+		if err != nil {
+			return err
+		}
 
-		config.Identities = append(config.Identities, Identity{
-			Name: models.InternalInfraConnectorIdentityName,
-		})
+		// load default connector identity
+		connector, err := s.loadIdentity(tx, Identity{Name: models.InternalInfraConnectorIdentityName}, provider)
+		if err != nil {
+			return err
+		}
 
-		config.Grants = append(config.Grants, Grant{
+		// load default connector grant
+		connectorGrant, err := s.loadGrant(tx, Grant{
 			Machine:  models.InternalInfraConnectorIdentityName,
 			Role:     models.InfraConnectorRole,
 			Resource: "infra",
 		})
+		if err != nil {
+			return err
+		}
 
-		if err := s.loadProviders(tx, config.Providers); err != nil {
+		if err := validator.New().Struct(s.options.Config); err != nil {
 			return err
 		}
 
-		// extract identities from grants and add them to identities
-		for _, g := range config.Grants {
-			switch {
-			case g.User != "":
-				config.Identities = append(config.Identities, Identity{Email: g.User})
-			case g.Machine != "":
-				config.Identities = append(config.Identities, Identity{Name: g.Machine})
+		// load and replace providers if specified in the config
+		if s.options.Config.Providers != nil {
```

Comment:
This is where we decide wether to "overwrite" or not – if users don't specify `providers:` in the server configuration then this won't run since the slice will be `nil` (as opposed to an empty slice – in which case this _will_ run and delete all providers as one may expect)